core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala

```diff
@@ -314,5 +315,60 @@ abstract class BaseConsumerTest extends IntegrationTestHarness with Logging {
 
     override def onComplete(offsets: util.Map[TopicPartition, OffsetAndMetadata], exception: Exception): Unit = count += 1
   }
+  protected class ConsumerAssignmentPoller(consumer: Consumer[Array[Byte], Array[Byte]]) extends ShutdownableThread("daemon-consumer-assignment", false)
+  {
+    private var partitionAssignment = scala.collection.mutable.Set[TopicPartition]()
+    private object lock
+
+    def consumerAssignment(): Set[TopicPartition] = {
+      lock.synchronized {
+        collection.immutable.Set(partitionAssignment.toArray: _*)
+      }
+    }
+
+    override def doWork(): Unit = {
+      consumer.poll(50)
+      lock.synchronized {
+        if (consumer.assignment() != partitionAssignment.asJava) {
+          partitionAssignment = consumer.assignment().asScala
+        }
+      }
+      Thread.sleep(100L)
+    }
+  }
+
+  /**
+   * Check whether partition assignment is valid
+   * Assumes partition assignment is valid iff
+   * 1. Every consumer got assigned at least one partition
+   * 2. Each partition is assigned to only one consumer
+   * 3. Every partition is assigned to one of the consumers
+   *
+   * @param assignments set of consumer assignments; one per each consumer
+   * @param partitions set of partitions that consumers subscribed to
+   * @return true if partition assignment is valid
+   */
+  def isPartitionAssignmentValid(assignments: Buffer[Set[TopicPartition]],
+                                 partitions: Set[TopicPartition]): Boolean = {
+    val allNonEmptyAssignments = assignments forall (assignment => assignment.size > 0)
+    if (!allNonEmptyAssignments) {
+      // at least one consumer got empty assignment
+      return false
+    }
+
+    // make sure that sum of all partitions to all consumers equals total number of partitions
+    val totalPartitionsInAssignments = (0 /: assignments) (_ + _.size)
```

Comment:
I don't know what fail-fast means yet, but I think my answer would be "for all cases". Basically I just want to check that each partition gets assigned to only one consumer and every partition get assigned to a consumer. Checking just the first two cases may return true in a case e.g. one partition gets assigned to two consumers and another partition does not get assigned at all. In this case, totalPartitionsInAssignments == partitions.size is true but the assignment is still invalid. So we are checking that merged set == partititions. Only checking merged set == partitions without checking counts may return valid assignment if there are duplicate partition assignments. So both second and third check is needed.